igvc_navigation/include/igvc_navigation/Graph.h

```diff
@@ -0,0 +1,301 @@
+  @param[in] s_prime diagonal node
+
+  @return diagonal traversal cost
+  */
+  float getC(const Node& s, const Node& s_prime);
+  /**
+  Returns traversal cost of node s and s', a non-diaginal (vertical or
+  horizontal) neighbor of s. Cost taken to be the maximum cost of
+  two cells neighboring the edge. If not occupied, TraversalCost is returned,
+   which is in units of (cost/distance).
+
+  @param[in] s reference node
+  @param[in] s_prime diagonal node
+
+  @return edge traversal cost
+  */
+  float getB(const Node& s, const Node& s_prime);
+  /**
+  Gets cost of traversing the grid cell while taking configuration space
+  into account
+
+  @param[in] ind index of cell to calculate ConfigurationSpace-corrected cost for
+  @return cost of traversing grid cell with ConfigurationSpace
+  */
+  float getValWithConfigurationSpace(const std::tuple<int, int>& ind);
+  /**
+  Gets nodes affected by updated cell value while taking into account ConfigurationSpace
+
+  @return list of reference nodes whose values may have been affected by
+  the updated cell cost
+  */
+  std::vector<Node> getUpdatedCellNodesWithConfigurationSpace();
+  /**
+  Get cost of traversing from a Node s to a neighboring node s_prime
+
+  @param[in] s node to get traversal cost through
+  @param[in] s_prime neighbor of s
+  @return cost of traversing from s to s_prime
+  */
+  float getTraversalCost(const Node& s, const Node& s_prime);
+  /**
+  Get cost of traversing from one continuous position on the graph p to another
+  continuous position p_prime
+  @param[in] p continuous position on the graph
+  @param[in] p_prime another continuous position on the graph
+  @return cost of traversing from p to p_prime
+  */
+  float getContinuousTraversalCost(const std::tuple<float, float>& p, const std::tuple<float, float>& p_prime);
+  /**
+  Gets minimum traversal cost from a node s to any neighboring node.
+
+  @param[in] s Node to get minimum traversal cost for
+  @return minimum traversal cost
+  */
+  float getMinTraversalCost(const Node& s);
+  /**
+  Calculates euclidian distance between the start node 'Start' and the
+  specified node s. This will be used as the focussing heuristic.
+
+  @param[in] s node to calculate euclidian distance to
+  @return euclidian distance between the start node and node s
+  */
+  float euclidianHeuristic(const Node& s);
+  /**
+  Same as above method but takes index of Node to calculate euclidian distance
+  from start for.
+
+  @param[in] ind (x,y) coordinates of node
+  @return euclidian distance to node from current start position
+  */
+  float euclidianHeuristic(const std::tuple<int, int>& ind);
+  /**
+  Gets Nodes around a cell whose occupancy value has changed while taking configuration
+  space into account.
+
+  @param[in] cellInd index of cell whose val has changed
+  @return list of nodes who might be affected by changed cell value
+  */
+  std::vector<Node> getNodesAroundCellWithConfigurationSpace(const Cell& cell);
```

Comment:
@barulicm Is returning a std::vector directly fine here? I feel like it's better to return through a reference parameter.

Reply:
This is something i'm always hesitant about. Any insight on this practice would be helpful.

Reply:
[This](https://stackoverflow.com/questions/15704565/efficient-way-to-return-a-stdvector-in-c) stack overflow post says that, in C++11, `std::vector<X> f();` is the preferred way to return a vector.

"With C++11, std::vector has move-semantics, which means the local vector declared in your function will be moved on return and in some cases even the move can be elided by the compiler."

Reply:
Even before C++11, you should never return a reference to a temporary variable. The vector returned here is created on demand, it's not something internal to the Graph object.

Even without move semantics, returning the plain vector type is the right thing to do.